### PR TITLE
[AIRFLOW-3284] Azure Batch AI Operator

### PR DESCRIPTION
**Pull request by elizabethhalper**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-3284\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3284
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-3284\], code changes always need a Jira issue.

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:
Add an operator to create an Azure Batch AI cluster. Azure Batch AI allows users to experiment and train deep learning and AI models in parallel at scale.

Operator supports creating a batch AI cluster by using any Marketplace image when given the publisher, offer, sku, and version.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

There are unit tests that test a successful execution of the operator and an execution that raises an Airflow exception.

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

**Comment by omusavi** [comment]
Looks like there are still merge markers in the code

**Comment by elizabethhalper** [comment]
@omusavi Sorry about that! I didn't realize they weren't deleted by auto-merging. I just went through and got rid of them.

File: airflow/contrib/example_dags/example_batchai_dag.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    'owner': 'airflow',
+    'start_date': airflow.utils.dates.days_ago(2)
+}
+
+dag = DAG(
+    dag_id='example_batchai_dag',
+    default_args=args,
+    schedule_interval="@daily")
+
+cmd = 'ls -l'
```
Comment (omusavi) [inline airflow/contrib/example_dags/example_batchai_dag.py]: Remove if this is unused

File: airflow/contrib/operators/azure_batchai_operator.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            image_reference = ImageReference(
+                publisher=self.publisher,
+                offer=self.offer,
+                sku=self.sku,
+                version=self.version)
+
+            vm_configuration = VirtualMachineConfiguration(
+                image_reference=image_reference)
+
+            username=os.environ['USERNAME'],
```
Comment (omusavi) [inline airflow/contrib/operators/azure_batchai_operator.py]: extraneous comma?

File: airflow/contrib/operators/azure_batchai_operator.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                messages = batch_ai_hook.get_messages(resource_group, name)
+                last_message_logged = self._log_last(messages, last_message_logged)
+                
+                if state == "Running":
+                    try:
+                        logs = batch_ai_hook.get_logs(resource_group, name)
+                        last_line_logged = self._log_last(logs, last_line_logged)
+                    except CloudError as err:
+                        self.log.exception("Exception while getting logs from "
+                                           "container instance, retrying...")
```
Comment (omusavi) [inline airflow/contrib/operators/azure_batchai_operator.py]: Should string be updated?

File: docs/integration.rst [hunk 4]
```diff
@@ -161,6 +161,25 @@ AzureDataLakeHook
...
+credentials for this principal can either be defined in the extra field `key_path`, as an 
+environment variable named `AZURE_AUTH_LOCATION`, or by providing a login/password and tenantId in extras.
+
+The AzureBatchAIHook requires a host/login/password to be defined in the connection.
+- :ref:`AzureBatchAIOperator` : Start/Monitor a new Azure Batch Workspace and Cluster.
+- :ref:`AzureBatchAIHook` : Wrapper around a single Batch AI workspace.
+
+AzureBatchAIOperator
+""""""""""""""""""""
+ .. autoclass:: airflow.contrib.operators.azure_container_instances_operator.AzureContainerInstancesOperator
```
Comment (omusavi) [inline docs/integration.rst]: These class names need to be updated

File: airflow/contrib/operators/azure_batchai_operator.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import os
+
+from time import sleep
+
+from airflow.contrib.hooks.azure_batchai_hook import (AzureBatchAIHook)
+                                                        # AzureContainerRegistryHook,
```
Comment (omusavi) [inline airflow/contrib/operators/azure_batchai_operator.py]: Remove commented out classes

File: airflow/contrib/hooks/azure_batchai_hook.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                                          workspace_name,
+                                          cluster_name)
+
+    def get_state_exitcode(self, resource_group, workspace_name, cluster_name):
+        response = self.connection.clusters.get(resource_group,
+                                                        workspace_name,
+                                                        cluster_name,
+                                                        raw=True).response.json()
+        cluster = response['properties']['cluster']
+        instance_view = cluster[0]['properties'].get('instanceView', {})
```
Comment (omusavi) [inline airflow/contrib/hooks/azure_batchai_hook.py]: can there be multiple clusters? Should we check to ensure there is at least 1? or >1 do we just do the first?

**Review by omusavi — CHANGES_REQUESTED** [review]
Looks good overall, just a few things I think we need before merging

File: tests/contrib/operators/test.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+
+# TEST SCRIPT BELOW
```
Comment (omusavi) [inline tests/contrib/operators/test.py]: looks like an accidental add :)

**Review by omusavi — APPROVED** [review]
Looks good!

**Comment by apraovjr** [comment]
@elizabethhalper I think title and description should match with apache guidelines for PR no?

**Comment by elizabethhalper** [comment]
@apraovjr I didn't think it would matter for our own repo but you're right, better to start now than have to change it later :)

**Comment by pchundi** [comment]
@omusavi: Can you merge the PR if it meets your expectations? 
We would like to test this for a POC.

**Comment by omusavi** [comment]
@pchundi I've already signed off but tests are not passing and are blocking merge

**Comment by elizabethhalper** [comment]
@pchundi I will merge as soon as the tests are passing

**Pull request merged**